sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java

```diff
@@ -410,6 +412,158 @@ public void testCompressedRead() throws Exception {
+
+    PrintStream writer = new PrintStream(out);
+    for (String line : lines) {
+      writer.println(line);
+      expected.add(line);
+    }
+    writer.close();
+    out.close();
+
+    // test with auto-detect ZIP based on extension.
+    Pipeline p = TestPipeline.create();
+
+    TextIO.Read.Bound<String> read = TextIO.Read.from(filename);
+    PCollection<String> output = p.apply(read);
+
+    PAssert.that(output).containsInAnyOrder(expected);
+    p.run();
+  }
+
+  /**
+   * Read a ZIP compressed empty file. We expect an exception.
+   */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testZipCompressedReadWithEmptyFile() throws Exception {
+    File tmpFile = tmpFolder.newFile();
+    String filename = tmpFile.getPath();
+
+    ZipOutputStream out = new ZipOutputStream(new FileOutputStream(tmpFile));
+
+    out.close();
+
+    Pipeline p = TestPipeline.create();
+
+    p.apply(TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP));
+    // expectedException.expect(IllegalArgumentException.class);
+    expectedException.expect(RuntimeException.class);
+    expectedException.expectMessage("ZIP file doesn't contain any entry");
+
+    p.run();
+  }
+
+  /**
+   * Read a ZIP compressed file containing an unique empty entry. We expect an empty PCollection.
+   */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testZipCompressedReadWithEmptyEntry() throws Exception {
+    File tmpFile = tmpFolder.newFile();
+    String filename = tmpFile.getPath();
+
+    ZipOutputStream out = new ZipOutputStream(new FileOutputStream(tmpFile));
+    out.putNextEntry(new ZipEntry("test.txt"));
+    out.closeEntry();
+
+    out.close();
+
+    Pipeline p = TestPipeline.create();
+
+    PCollection<String> output = p.apply(TextIO.Read.from(filename).withCompressionType
+        (CompressionType.ZIP));
+    PAssert.that(output).empty();
+
+    p.run();
+  }
+
+  /**
+   * Read a ZIP compressed file with multiple entries. Only the first entry is actually read.
+   * We expect an empty PCollection.
+   */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testZipCompressedReadWithMultiEntriesFile() throws Exception {
+    File tmpFile = tmpFolder.newFile();
+    String filename = tmpFile.getPath();
+
+    ZipOutputStream out = new ZipOutputStream(new FileOutputStream(tmpFile));
+
+    out.putNextEntry(new ZipEntry("first.txt"));
+    out.closeEntry();
+    out.putNextEntry(new ZipEntry("second.txt"));
+    PrintStream writer = new PrintStream(out);
+    writer.println("This is a test");
+    out.closeEntry();
+    out.putNextEntry(new ZipEntry("third.txt"));
+    writer.println("This is a test");
+    out.closeEntry();
+    writer.close();
+
+    out.close();
+
+    Pipeline p = TestPipeline.create();
+
+    TextIO.Read.Bound<String> read =
+        TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP);
+    PCollection<String> output = p.apply(read);
+
+    PAssert.that(output).empty();
+    p.run();
```

Comment:
I'd really like to find a way that this case either concats all the files or throws an exception. The current behavior is effectively silent data loss, the worst possible case!

One way to handle this could be a utility InputStream class that wraps `ZipInputStream` and under-the-hood concats all the different entry streams. This is probably the base case.

A second possibility is that once the input stream hits EOF, we check for a next entry and only then throw an exception. But this is less desirable as we don't fail until pretty late.

Can you look into it?